File: firebase/functions/lib/utils/mux_client.dart
import 'dart:convert' as convert;
import 'dart:convert';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import 'package:node_http/node_http.dart' as http;

final muxApi = MuxApi();

class MuxApi {
  static final _muxTokenId = functions.config.get('mux.token_id') as String;
  static final _muxSecret = functions.config.get('mux.secret') as String;

  static String get _auth => '$_muxTokenId:$_muxSecret';

  final Map<String, String> _headers = {
    'Authorization': 'Basic ${base64.encode(_auth.codeUnits)}',
    'Content-Type': 'application/json',
  };

  Future<dynamic> createLiveStream() async {
    final response = await http.post(
      Uri.parse('https://api.mux.com/video/v1/live-streams'),
      headers: _headers,
      body: convert.json.encode(
        {
          'playback_policy': 'public',
          'reduced_latency': true,
          'new_asset_settings': {
            'playback_policy': 'public',
          },
        },
      ),
    );

    _verifyResponse(response);

    return convert.json.decode(response.body)['data'];
  }

  void _verifyResponse(http.Response response) {
    if (response.statusCode < 200 || response.statusCode > 299) {
      print('Error during mux call:');
      print(response.statusCode);
      print(response.body);
      throw Exception(response.body);
    }
  }
}

File: firebase/functions/src/events/live_meetings/mux_client.ts
import * as functions from 'firebase-functions';
import * as https from 'https';
import * as http from 'http';

function makeRequest(options: http.RequestOptions, body?: string): Promise<{ statusCode: number; body: string }> {
  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => resolve({ statusCode: res.statusCode ?? 0, body: data }));
    });
    req.on('error', reject);
    if (body) req.write(body);
    req.end();
  });
}

export const muxApi = new class MuxApi {
  private get _muxTokenId(): string { return functions.config().mux?.token_id as string; }
  private get _muxSecret(): string { return functions.config().mux?.secret as string; }
  private get _auth(): string { return `${this._muxTokenId}:${this._muxSecret}`; }

  private get _headers(): Record<string, string> {
    return {
      Authorization: `Basic ${Buffer.from(this._auth).toString('base64')}`,
      'Content-Type': 'application/json',
    };
  }

  async createLiveStream(): Promise<Record<string, unknown>> {
    const body = JSON.stringify({
      playback_policy: 'public',
      reduced_latency: true,
      new_asset_settings: { playback_policy: 'public' },
    });
    const result = await makeRequest({
      method: 'POST',
      hostname: 'api.mux.com',
      path: '/video/v1/live-streams',
      headers: { ...this._headers, 'Content-Length': Buffer.byteLength(body).toString() },
    }, body);
    this._verifyResponse(result);
    return (JSON.parse(result.body) as { data: Record<string, unknown> }).data;
  }

  private _verifyResponse(response: { statusCode: number; body: string }): void {
    if (response.statusCode < 200 || response.statusCode > 299) {
      console.error('Error during mux call:', response.statusCode, response.body);
      throw new Error(response.body);
    }
  }
}();
